resolwe/storage/manager.py

```diff
@@ -0,0 +1,188 @@
+        self.file_storage = file_storage
+
+    def _get_rules(self, connector_name: str, _type: str) -> dict:
+        """Get rules for the given location. Valid types of rules are "copy"
+        and "delete".
+        """
+        location_settings = settings.STORAGE_CONNECTORS.get(connector_name, {})
+        return location_settings.get("config", {}).get(_type, {})
+
+    def copy(self) -> List[str]:
+        """Decide if the file storage object should be copied to more storage
+        locations.
+
+        :return: a list connector names where data should be copied to.
+        :rtype: list[str]
+        """
+        return [
+            connector_name
+            for connector_name in settings.STORAGE_CONNECTORS
+            if self._should_copy_to(connector_name)
+        ]
+
+    def _should_copy_to(self, connector_name: str) -> bool:
+        """Decide if data should be copied to the given storage connector
+        accoring to the rules.
+
+        :param connector_name: name of the storage connector.
+        :type connector_name: str
+
+        :return: True when data should be copied to the location managed by
+            connector connector_name.
+        """
+        copy_rules = self._get_rules(connector_name, "copy")
+
+        rule_results = {}
+
+        rule_results["has_rules"] = len(copy_rules) > 0
+
+        rule_results["copy_not_present"] = not self.file_storage.has_storage_location(
+            connector_name
+        )
+
+        if "delay" in copy_rules:
+            move_delay = timedelta(days=copy_rules["delay"])
+            current_delay = now() - self.file_storage.created
+            rule_results["delay"] = current_delay >= move_delay
+
+        return all(rule_results.values())
+
+    def delete(self) -> List[StorageLocation]:
+        """Decide which StorageLocation for the FileStorage object should be
+        removed.
+
+        :return: a list of registered storage connector names from which the
+            data should be removed.
+        :rtype: List[StorageLocation]
+        """
+        return [
+            storage_location
+            for storage_location in self.file_storage.storage_locations.all()
+            if self._should_delete(storage_location)
+        ]
+
+    def _should_delete(self, storage_location: StorageLocation):
+        """Decide if the given StorageLocation object should be deleted.
+
+        :return: True if storage_location should be deleted, False otherwise.
+        :rtype: bool
+        """
+        rules = self._get_rules(storage_location.connector_name, "delete")
+        rule_results = {}
+        rule_results["has_rules"] = len(rules) > 0
+        rule_results["not_locked"] = not storage_location.locked
+
+        if "min_other_copies" in rules:
+            n_r_locs = rules["min_other_copies"]
+            n_c_locs = self.file_storage.storage_locations.count()
+            rule_results["has_copies"] = n_c_locs - 1 >= n_r_locs
```

Comment:
Maybe I should process delete rules before copy rules and keep track of which storage locations are to be deleted?

Reply:
I think that would make the situation even worse because even more rules can change in the meantime.

What if this method only returns the first location to delete, and it is called multiple times?

Reply:
Done.